kernel/mem_domain.c

```diff
@@ -0,0 +1,294 @@
+		/* A zero-sized partition denotes it's a free partition */
+		if (domain->partitions[p_idx].size == 0) {
+			break;
+		}
+	}
+
+	/* No free partition */
+	if (p_idx == max_partitions) {
+		irq_unlock(key);
+		return -ENOMEM;
+	}
+
+	domain->partitions[p_idx].start = part->start;
+	domain->partitions[p_idx].size = part->size;
+	domain->partitions[p_idx].attr = part->attr;
+
+	domain->num_partitions++;
+
+	SYS_DLIST_FOR_EACH_NODE_SAFE(&domain->mem_domain_q, node, next_node) {
+		struct k_thread *thread =
+			CONTAINER_OF(node, struct k_thread, mem_domain_info);
+
+		/*
+		 * Call arch specific API to configure a single
+		 * memory partition.
+		 */
+		arch_configure_mem_partition(thread, p_idx);
+	}
+
+	irq_unlock(key);
+
+	return 0;
+}
+
+int k_mem_domain_remove_partition(struct k_mem_domain *domain,
+				  struct k_mem_partition *part)
+{
+	int p_idx;
+	unsigned int key;
+	sys_dnode_t *node, *next_node;
+
+	if (!domain || !part) {
+		return -EINVAL;
+	}
+
+	key = irq_lock();
+
+	/* find a partition that matches the given start and size */
+	for (p_idx = 0; p_idx < max_partitions; p_idx++) {
+		if (domain->partitions[p_idx].start == part->start &&
+		    domain->partitions[p_idx].size == part->size) {
+			break;
+		}
+	}
+
+	/* Not found */
+	if (p_idx == max_partitions) {
+		irq_unlock(key);
+		return -EINVAL;
+	}
+
+	SYS_DLIST_FOR_EACH_NODE_SAFE(&domain->mem_domain_q, node, next_node) {
+		struct k_thread *thread =
+			CONTAINER_OF(node, struct k_thread, mem_domain_info);
+
+		/*
+		 * Call arch specific API to disable a single
+		 * memory partition.
+		 */
+		arch_disable_mem_partition(thread, p_idx);
+	}
+
+	domain->partitions[p_idx].start = 0;
+	domain->partitions[p_idx].size = 0;
+	domain->partitions[p_idx].attr = 0;
+
+	domain->num_partitions--;
+
+	irq_unlock(key);
+
+	return 0;
+}
+
+int k_mem_domain_add_thread(struct k_mem_domain *domain, k_tid_t thread)
+{
+	unsigned int key;
+
+	if (!domain || !thread || thread->mem_domain_info.mem_domain) {
+		return -EINVAL;
+	}
+
+	key = irq_lock();
+
+	sys_dlist_append(&domain->mem_domain_q,
+			 &thread->mem_domain_info.mem_domain_q_node);
+	thread->mem_domain_info.mem_domain = domain;
+
+	/*
+	 * Configure memory domain immediately if the thread that is added
+	 * into the domain is the current thread
+	 */
+	if (thread == _current) {
+		arch_configure_mem_domain(_current);
+	}
+
+	irq_unlock(key);
+
+	return 0;
+}
+
+int k_mem_domain_remove_thread(k_tid_t thread)
+{
+	unsigned int key;
+
+	if (thread == NULL || !thread->mem_domain_info.mem_domain) {
+		return -EINVAL;
+	}
+
+	key = irq_lock();
+
+	/*
+	 * Call arch specific API to disable thread's memory
+	 * partitions.
+	 */
+	arch_disable_all_mem_partitions(thread);
+
+	sys_dlist_remove(&thread->mem_domain_info.mem_domain_q_node);
+	thread->mem_domain_info.mem_domain = NULL;
+
+	irq_unlock(key);
+
+	return 0;
+}
+
+static int init_mem_domain_module(struct device *arg)
+{
+	ARG_UNUSED(arg);
+
+	max_partitions = arch_get_max_partitions();
```

Comment:
Is this something we can know at build time, rather than have to derive at runtime?

Reply:
It is possible to know it at build time if soc-specific codes could provide the information of its maximum supported regions.
Because the maximum supported regions might be different from each SoC and this information might not be found in their user manual, it is a little bit annoying to add this information into the each SoC that we currently have.
So I deicde to derive it at runtime finally.

Reply:
ok, we can leave this as-is